toolchain/build-toolchain.sh

```diff
@@ -30,76 +45,151 @@ GCC=gcc-9.3.0
+    export PATH
 
-cd "${BUILD_ROOT}"
+    log "installing kernel headers"
+    make -C "$SCRIPT_DIR/../../phoenix-rtos-kernel" TARGET="$PHOENIX_TARGET" install-headers
 
-log "downloading GCC dependencies"
-(cd "$GCC" && ./contrib/download_prerequisites)
+    # FIXME: libphoenix should be installed for all supported multilib target variants
+    log "installing libphoenix"
+    make -C "$SCRIPT_DIR/../../libphoenix" TARGET="$PHOENIX_TARGET" clean install
 
+    PATH="$OLDPATH"
+}
 
-log "patching ${GCC}"
-for patchfile in "${GCC}"-*.patch; do
-    if [ ! -f "${GCC}/$patchfile.applied" ]; then
-        patch -d "${GCC}" -p1 < "$patchfile"
-        touch "${GCC}/$patchfile.applied"
-    fi
-done
+build_gcc_stage2() {
+    pushd "$BUILD_ROOT/${GCC}/build" > /dev/null
 
-log "building GCC"
-rm -rf "${GCC}/build"
-mkdir -p "${GCC}/build"
-cd "${GCC}/build"
+    # (hackish) instead of reconfiguring and rebuilding whole gcc
+    # just force rebuilding internal includes (and fixincludes)
+    # remove stamp file for internal headers generation
+    rm gcc/stmp-int-hdrs
 
-if [[ "$TARGET" = *"arm-" ]]; then
-    GCC_CONFIG_PARAMS="--with-abi=aapcs"
-fi
+    log "building GCC (stage2)"
+    make all-gcc all-target-libgcc
+
+    log "installing GCC (stage2)"
+    make install-gcc install-target-libgcc
+
+    # TODO: build libstdc++ -> for now it fails because of libphoenix headers
+    # (no compilation possible with no flags - #error "unsupported architecture")
+    # also: not sure if we shouldn't install libphoenix headers in ${sysroot}/include instead of ${sysroot}/usr/include:
+    # - this path is hardcoded while building libstdc++ using stage1 compiler
+    # - newlib headers in arm-none-eabi-gcc are installed that way
+    # reference: https://github.com/xpack-dev-tools/arm-gcc-original-scripts/blob/master/build-toolchain.sh
+
+    popd > /dev/null
+}
+
+strip_binaries() {
+    log "stripping binaries"
+    for bin in "$TOOLCHAIN_PREFIX"/bin/*; do
+        strip "$bin"
+    done
+
+    # NOTE: we could also strip target libraries, but let's leave debug sections for ease of development
+}
+
+
+### MAIN ###
 
-# GCC compiler options
-# --with-sysroot -> cross-compiler sysroot
-# --with-newlib -> do note generate standard library includes by fixincludes, do not include _eprintf in libgcc
-# --disable-libssp -> stack smashing protector library disabled
-# --disable-nls -> all compiler messages will be in english
+# comment out some parts if You need "incremental build" for testing
 
-../configure --target="${TARGET}" --prefix="${TOOLCHAIN_PREFIX}" \
-             --with-sysroot="${TOOLCHAIN_PREFIX}/${TARGET}" \
-             --enable-languages=c --with-newlib \
-             --disable-libssp --disable-nls $GCC_CONFIG_PARAMS
+download;
+build_binutils;
+build_gcc_stage1;
 
-make ${MAKEOPTS} all-gcc all-target-libgcc
+build_libc;
+build_gcc_stage2;
 
-log "installing GCC"
-make install-gcc install-target-libgcc
+strip_binaries;
```

Comment:
Semicolons at the end of the line are not needed.

Reply:
Maybe add a message at the end of the file
```
log "Toolchain for $TARGET instaled in directory $TOOLCHAIN_PREFIX
```

Reply:
- semicolons - I know, but I prefer the syntax
- added message at successful script end
- no coding style for shell scripts, rule of thumb is to keep current style intact